Show new files as file, not diff

gerhardol opened the pull request:
Fixes #5329 
Follow up to #7824

## Proposed changes

Always show new files as files, not as a patch

Note: A sideeffect for this is that it will not be possible to cherry-pick/revert lines for new files (except FormCommit, as is now).
This is planned for revDiff artificial commits in #7825 , not for other commits or forms right now.

## Screenshots <!-- Remove this section if PR does not change UI -->

### Before

![image](https://user-images.githubusercontent.com/6248932/75617531-b206ce80-5b60-11ea-85df-cc21e1871e8e.png)

### After

![image](https://user-images.githubusercontent.com/6248932/75617504-52a8be80-5b60-11ea-905e-df636506a877.png)

![image](https://user-images.githubusercontent.com/6248932/75617513-6bb16f80-5b60-11ea-8537-6bb824d9d1c4.png)

## Test methodology 

Manual

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

gerhardol commented:
If this is merged, FileViewer:ViewCurrentChanges() could be replaced with GitUIExtensions.cs: ViewChangesAsync() to unify the handling further.
#7825 is needed (should be done in that PR.)

mstv commented:
Sorry, I am too busy for the code review this week.
But a question and a wish arose:
- How are deleted files displayed?
- A header with the file name would be very useful for me, i.e.
  "+++ b/path/file.ext" or "--- a/path/file.ext", respectively.

gerhardol commented:
> How are deleted files displayed?

As a patch.
Changing this a little more tricky, especially if line patching is to be supported.

> A header with the file name would be very useful for me, i.e.
    "+++ b/path/file.ext" or "--- a/path/file.ext", respectively.

That will interfere with line patching, not so easy at it seems

To change the file presentation like this for normal diffs is something discussed before and should be considered.

mstv commented:
> > How are deleted files displayed?

> As a patch.

Good, so a delete file looks different than an added one.

> header ... will interfere with line patching

OK, no issue for me. Line patching has high precedence.

Though a quite late question, asked for the last time: What is the benefit of displaying new files not as diff? It seems to require extra handling, e.g. for line patching.

gerhardol commented:
> Though a quite late question, asked for the last time: What is the benefit of displaying new files not as diff? It seems to require extra handling, e.g. for line patching.

The benefit from the previous was to align FormCommit and RevisionDiff, as well as setting FilePreamble in FileViewer, required for patching (there are other ways to fix this though).
#7824 was therefore required for #7825, this is just a follow up.

mstv commented on GitUI/Editor/FileViewer.cs:
```diff
@@ -530,19 +530,19 @@ public void ViewCurrentChanges(GitItemStatus item, bool isStaged, [CanBeNull] Ac
+                    if (!item.IsSubmodule)
                     {
-                        await ViewFileAsync(item.Name, item.IsSubmodule, openWithDifftool);
-                        SetVisibilityDiffContextMenuStaging();
```
Can the call of `SetVisibilityDiffContextMenuStaging` really be removed?
If yes, the method itself should be removed. Or will it be reused in a follow-up PR?

Slightly off-topic: FileStatusList context menu:
I tried to `cherry pick changes` from a new file in an unmerged branch (like "bugfix/i6723-filename-space-2") without effect.
I guess this menu item is not applicable for new files. The same applies to `Reset file(s) to | Selected`.

gerhardol replied:
> Can the call of SetVisibilityDiffContextMenuStaging really be removed?
If yes, the method itself should be removed. Or will it be reused in a follow-up PR?

SetVisibilityDiffContextMenuStaging is not removed, just moved around.
However #7825 removes SetVisibilityDiffContextMenuStaging  and a follow up to that will remove ViewCurrentChanges() and let all "patchviewers" use ViewChangesAsync()
The unification is why I prefer to look over this after #7825

mstv replied:
Sounds good.

gerhardol commented:
@mstv any comments?
I have several changes depending on this, so I will initiate a merge

For cherry picking patches, I plan to clean this up but likely not add patching unless Artificial commits are selected. Such patches fail often anyway.

gerhardol commented on GitUI/Editor/FileViewer.cs:
```diff
@@ -530,19 +530,19 @@ public void ViewCurrentChanges(GitItemStatus item, bool isStaged, [CanBeNull] Ac
+                            await ViewPatchAsync(item.Name, patch?.Text ?? "", openWithDifftool, isText: false);
+                        }
                     }
                     else
```
Pushed a variant of this - plan to remove anyway

gerhardol commented:
> Slightly off-topic: Removed image files are displayed as empty (also before this PR).

A new and removed image would either be shown the same way or a text is shown. (Overcourse: a link to show the removed image.)

gerhardol commented:
Added a commit to handle unstage/revert for Index in FormCommit (this is actually the same as after other changes).

RevDiff for WorkTree/Index will be added to  #7825.
For other combinations I do not know yet, that depends on other reasons too, like if a file is staged or not. That is not changed from the current behavior, just that the menu item will not be visible.
Will probably leave that out (at least for some time), not important anyway.

gerhardol commented:
Rebased and squashed, added a variable for readability

gerhardol commented:
@msftbot merge in 24 hours
